### PR TITLE
[SPARK-25947][SQL] Reduce memory usage in ShuffleExchangeExec by selecting only the sort columns

### PR DESCRIPTION
**Pull request by mu5358271**
## What changes were proposed in this pull request?

When sorting rows, ShuffleExchangeExec uses the entire row instead of just the columns references in SortOrder to create the RangePartitioner. This causes the RangePartitioner to sample entire rows to create rangeBounds and can cause OOM issues on the driver when rows contain large fields.

This change creates a projection and only use columns involved in the SortOrder for the RangePartitioner

## How was this patch tested?

Existing tests in spark-sql. 

Plus

Started a local spark-shell with a small spark.driver.maxResultSize:

```
spark-shell --master 'local[16]' --conf spark.driver.maxResultSize=128M --driver-memory 4g
```

and ran the following script:

```
import com.google.common.io.Files
import org.apache.spark.SparkContext
import org.apache.spark.sql.SparkSession

import scala.util.Random

@transient val sc = SparkContext.getOrCreate()
@transient val spark = SparkSession.builder().getOrCreate()

import spark.implicits._

val path = Files.createTempDir().toString

// this creates a dataset with 1024 entries, each 1MB in size, across 16 partitions
sc.parallelize(0 until (1 << 10), sc.defaultParallelism).
  map(_ => Array.fill(1 << 18)(Random.nextInt)).
  toDS.
  write.mode("overwrite").parquet(path)

spark.read.parquet(path).
  orderBy('value (0)).
  write.mode("overwrite").parquet(s"$path-sorted")

spark.read.parquet(s"$path-sorted").show
```
execution would fail when initializing RangePartitioner without this change.
execution succeeds and generates a correctly sorted dataset with this change.

Please review http://spark.apache.org/contributing.html before opening a pull request.

### CODE REVIEW

**Comment by mu5358271** [comment]
cc @cloud-fan @gatorsmile @hvanhovell

File: sql/core/src/main/scala/org/apache/spark/sql/execution/exchange/ShuffleExchangeExec.scala [hunk 1]
```diff
@@ -214,13 +214,24 @@ object ShuffleExchangeExec {
...
+            UnsafeProjection.create(sortingExpressions.map(_.child), outputAttributes)
           val mutablePair = new MutablePair[InternalRow, Null]()
-          iter.map(row => mutablePair.update(row.copy(), null))
+          // Internally, RangePartitioner runs a job on the RDD that samples keys to compute
+          // partition bounds. To get accurate samples, we need to copy the mutable keys.
+          iter.map(row => mutablePair.update(projection(row).copy(), null))
         }
-        implicit val ordering = new LazilyGeneratedOrdering(sortingExpressions, outputAttributes)
+        // Construct ordering on extracted sort key.
+        val orderingAttributes =
```
Comment (hvanhovell) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/exchange/ShuffleExchangeExec.scala]: This is a bit clunky IMO. Can we do this instead:
```scala
val orderingAttributes = sortingExpressions.zipWithIndex.map { case (ord, i) =>
  ord.copy(child = BoundReference(i, ord.dataType, ord.nullable))
}
```

Comment (mu5358271) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/exchange/ShuffleExchangeExec.scala]: you are right. this is much better. changed

**Comment by cloud-fan** [comment]
ok to test

File: sql/core/src/main/scala/org/apache/spark/sql/execution/exchange/ShuffleExchangeExec.scala [hunk 2]
```diff
@@ -214,13 +214,22 @@ object ShuffleExchangeExec {
...
+          // partition bounds. To get accurate samples, we need to copy the mutable keys.
+          iter.map(row => mutablePair.update(projection(row).copy(), null))
         }
-        implicit val ordering = new LazilyGeneratedOrdering(sortingExpressions, outputAttributes)
+        // Construct ordering on extracted sort key.
+        val orderingAttributes = sortingExpressions.zipWithIndex.map { case (ord, i) =>
+          ord.copy(child = BoundReference(i, ord.dataType, ord.nullable))
+        }
+        implicit val ordering: Ordering[InternalRow] =
+          new LazilyGeneratedOrdering(orderingAttributes)
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/exchange/ShuffleExchangeExec.scala]: style nit: this can be merged to the previous line

**Comment by cloud-fan** [comment]
do you have some benchmark numbers?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/exchange/ShuffleExchangeExec.scala (hunk 2 above)
Comment (mu5358271) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/exchange/ShuffleExchangeExec.scala]: this line would actually exceed the 100 character per line limit by 2 characters if I keep the ": Ordering[InternalRow]" type info for the implicit value. I can remove the type info though. Is that what you are suggesting?

Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/exchange/ShuffleExchangeExec.scala]: yea, let's follow the previous style: https://github.com/apache/spark/pull/22961/files#diff-3ceee31a3da1b7c71dddd32f666126fbL223

**Comment by mu5358271** [comment]
Did some performance evaluation on a 1G test dataset on a small cluster with the following script:

```
import java.util.UUID

import org.apache.spark.SparkContext
import org.apache.spark.sql.SparkSession
import org.apache.spark.sql.functions.udf

import scala.util.{Random, Try}

@transient val sc = SparkContext.getOrCreate()
@transient val spark = SparkSession.builder().getOrCreate()

import spark.implicits._

val totalSize = 28 // 1G total

val longSize = 6 // 256 Byte records

val wideSize = 13 // 32KB records

sc.parallelize(0 until (1 << (totalSize - longSize)), 200).
  map(_ => Array.fill(1 << longSize)(Random.nextInt)).
  toDS.
  write.mode("overwrite").parquet("long")

sc.parallelize(0 until (1 << (totalSize - wideSize)), 200).
  map(_ => Array.fill(1 << wideSize)(Random.nextInt)).
  toDS.
  write.mode("overwrite").parquet("wide")

val expensiveOrdering = udf((vs: Seq[Int]) => vs.foldLeft(0L)(_ + _))

for {
  format <- Seq("wide", "long")
  expensive <- Seq(true, false)
  trial <- 0 until 10
} yield {
  val time =
    Try({
      val start = System.currentTimeMillis()
      spark.read.parquet(format).orderBy(if (expensive) expensiveOrdering('value) else 'value (0)).write.parquet(s"$format-${UUID.randomUUID}")
      System.currentTimeMillis() - start
    }).toOption
  (format, expensive, trial, time)
}
```

scenarios:
- after: with this change and using default 1g spark.driver.maxResultSize 
- before : without this change and using default 1g spark.driver.maxResultSize
- before + : without this change and increase spark.driver.maxResultSize from default 1g to 4g.

no value means evaluation failed.

ordering | format | scenario | avg (ms) | 0 | 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
cheap | long | after | 23435.2 | 16259 | 31116 | 24585 | 21104 | 26732 | 15863 | 23716 | 25672 | 28313 | 20992
cheap | long | before | 24087.9 | 26391 | 24483 | 28731 | 24995 | 18151 | 27224 | 25278 | 16526 | 24290 | 24810
cheap | long | before+ | 21538.5 | 22336 | 31748 | 17915 | 21733 | 16393 | 20415 | 23558 | 21403 | 22264 | 17620
cheap | wide | after | 25028.7 | 26401 | 21526 | 27118 | 22763 | 41360 | 14608 | 22935 | 28918 | 21304 | 23354
cheap | wide | before |   |   |   |   |   |   |   |   |   |   |  
cheap | wide | before+ | 33324.1 | 42077 | 32455 | 38926 | 31055 | 30729 | 30532 | 30121 | 30127 | 30357 | 36862
expensive | long | after | 24989.2 | 22967 | 22490 | 22365 | 27159 | 23944 | 25401 | 22834 | 26073 | 28212 | 28447
expensive | long | before | 33553.1 | 30019 | 33404 | 32004 | 33547 | 35282 | 34149 | 33365 | 30934 | 36945 | 35882
expensive | long | before+ | 32839.4 | 32572 | 35354 | 32635 | 33385 | 32063 | 33350 | 35472 | 31771 | 31261 | 30531
expensive | wide | after | 26740.2 | 39559 | 30116 | 22777 | 24766 | 21391 | 22470 | 31302 | 18392 | 35768 | 20861
expensive | wide | before |   |   |   |   |   |   |   |   |   |   |  
expensive | wide | before+ | 254233.4 | 356997 | 309464 | 281589 | 226232 | 223588 | 224295 | 238064 | 226036 | 230633 | 225436


- the suggested change has roughly the same performance as before when the dataset has small rows and the ordering evaluation is cheap.
- it reduces runtime when the ordering evaluation is expensive by distributing ordering evaluation across the cluster.
- it reduces driver memory usage and helps job complete successfully when rows are large by reducing the size of data collected to the driver

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Pull request closed without merging**